src/topology/sheaves/presheaf.lean

```diff
@@ -133,21 +139,108 @@ def pushforward_map {X Y : Top.{v}} (f : X ⟶ Y) {ℱ 𝒢 : X.presheaf C} (α
 { app := λ U, α.app _,
   naturality' := λ U V i, by { erw α.naturality, refl, } }
 
+open category_theory.limits
+section pullback
+variable [has_colimits C]
+noncomputable theory
+
+/--
+Pullback a presheaf on `Y` along a continuous map `f : X ⟶ Y`, obtaining a presheaf on `X`.
+
+This is defined in terms of left Kan extensions, which is just a fancy way of saying
+"take the colimits over the open sets whose preimage contains U".
+-/
+@[simps]
+def pullback_obj {X Y : Top.{v}} (f : X ⟶ Y) (ℱ : Y.presheaf C) : X.presheaf C :=
+(Lan (opens.map f).op).obj ℱ
+
+/-- Pulling back along continuous maps is functorial. -/
+def pullback_map {X Y : Top.{v}} (f : X ⟶ Y) {ℱ 𝒢 : Y.presheaf C} (α : ℱ ⟶ 𝒢) :
+  pullback_obj f ℱ ⟶ pullback_obj f 𝒢 :=
+(Lan (opens.map f).op).map α
+
+/-- If `f '' U` is open, then `f⁻¹ℱ U ≅ ℱ (f '' U)`.  -/
+@[simps]
+def pullback_obj_obj_of_image_open {X Y : Top.{v}} (f : X ⟶ Y) (ℱ : Y.presheaf C) (U : opens X)
+  (H : is_open (f '' U)) : (pullback_obj f ℱ).obj (op U) ≅ ℱ.obj (op ⟨_, H⟩) :=
+begin
+  let x : costructured_arrow (opens.map f).op (op U) :=
+  { left := op ⟨f '' U, H⟩,
+    hom := ((@hom_of_le _ _ _ ((opens.map f).obj ⟨_, H⟩) (set.image_preimage.le_u_l _)).op :
+    op ((opens.map f).obj (⟨⇑f '' ↑U, H⟩)) ⟶ op U) },
+  have hx : is_terminal x :=
+  { lift := λ s, by { fapply costructured_arrow.hom_mk,
+    change op (unop _) ⟶ op (⟨_, H⟩ : opens _),
+    refine (hom_of_le _).op,
+    exact (set.image_subset f s.X.hom.unop.le).trans (set.image_preimage.l_u_le ↑(unop s.X.left)),
+    simp } },
+  exact is_colimit.cocone_point_unique_up_to_iso
+    (colimit.is_colimit _)
+    (colimit_of_diagram_terminal hx _),
+end
+
+namespace pullback
+variables {X Y : Top.{v}} (ℱ : Y.presheaf C)
+
+local attribute [reassoc] colimit.pre_desc
+
+/-- The pullback along the identity is isomorphic to the original presheaf. -/
+def id : pullback_obj (𝟙 _) ℱ ≅ ℱ :=
+nat_iso.of_components
+  (λ U, pullback_obj_obj_of_image_open (𝟙 _) ℱ (unop U) (by simpa using U.unop.2) ≪≫
+    ℱ.map_iso (eq_to_iso (by simp)))
+  (λ U V i, by {
```

Comment:
The style guide reserves `by { .. }` for proofs that are ≤ 2 lines. Please use `begin .. end` blocks for longer proofs